src/components/cylc/tree/InfiniteTree.vue

```diff
@@ -0,0 +1,202 @@
+  props: {
+    treeData: {
+      type: [Array, Object],
+      default: () => {
+        return []
+      }
+    },
+    autoOpen: {
+      type: Boolean,
+      default: false
+    },
+    selectable: {
+      type: Boolean,
+      default: false
+    },
+    tabIndex: {
+      type: Number,
+      default: 1
+    },
+    className: {
+      type: String,
+      default: 'scroll-box'
+    },
+    rowHeight: {
+      type: Number,
+      default: 32
+    },
+    loadNodes: {
+      type: Function,
+      default: () => {}
+    },
+    shouldSelectNode: {
+      type: Function,
+      default: () => {}
+    },
+    shouldLoadNodes: {
+      type: Function,
+      default: () => {}
+    },
+    // Callback invoked before updating the tree.
+    onContentWillUpdate: {
+      type: Function,
+      default: null
+    },
+    // Callback invoked when the tree is updated.
+    onContentDidUpdate: {
+      type: Function,
+      default: null
+    },
+    // Callback invoked when a node is opened.
+    onOpenNode: {
+      type: Function,
+      default: null
+    },
+    // Callback invoked when a node is closed.
+    onCloseNode: {
+      type: Function,
+      default: null
+    },
+    // Callback invoked when a node is selected or deselected.
+    onSelectNode: {
+      type: Function,
+      default: null
+    },
+    // Callback invoked before opening a node.
+    onWillOpenNode: {
+      type: Function,
+      default: null
+    },
+    // Callback invoked before closing a node.
+    onWillCloseNode: {
+      type: Function,
+      default: null
+    },
+    // Callback invoked before selecting or deselecting a node.
+    onWillSelectNode: {
+      type: Function,
+      default: null
+    },
+    onKeyUp: {
+      type: Function,
+      default: null
+    },
+    onKeyDown: {
+      type: Function,
+      default: null
+    },
+    onMouseLeave: {
+      type: Function,
+      default: null
+    },
+    onMouseEnter: {
+      type: Function,
+      default: null
+    }
+  },
+  inheritAttrs: false,
+  data () {
+    return {
+      tree: {
+        nodes: []
+      },
+      nodes: [],
+      eventHandlers: {
+        onContentWillUpdate: null,
+        onContentDidUpdate: null,
+        onOpenNode: null,
+        onCloseNode: null,
+        onSelectNode: null,
+        onWillOpenNode: null,
+        onWillCloseNode: null,
+        onWillSelectNode: null,
+        onKeyUp: null,
+        onKeyDown: null,
+        onMouseEnter: null,
+        onMouseLeave: null
+      }
+    }
+  },
+  watch: {
+    treeData: {
+      handler (newValue) {
+        this.tree.loadData(newValue)
+      }
+    }
+  },
+  mounted () {
+    this.tree = new InfiniteTree({
+      el: this.$refs.tree,
+      ...this.$props,
+      data: this.$props.treeData
+    })
+
+    this.tree.update = () => {
+      this.tree.emit('contentWillUpdate')
+      this.nodes = this.tree.nodes
+      this.$nextTick(function () {
+        this.tree.emit('contentDidUpdate')
+      })
```

Comment:
This emits a `contendDidUpdate` event on the next "tick", do we know the content actually updated? I don't know how these ticks work.

Reply:
These ticks are used by Vue to wait for a DOM update.

In the case of the `InfiniteTree`, we have the widget in the DOM. Whenever we manipulate the DOM in Vue, we need to consider whether we are ready to run something, or if we are still with a Virtual DOM node - a VNode.

Later, Vue will translate the VNode to the DOM node. When that happens, Vue gives us a change to execute something - on the next tick.

Not sure if a good explanation, as the weekends gets closer, may English gets worse :smile:

Reply:
Good explanation. So a DOM update will only ever take take one "tick" therefore we can guarentee that the update will be complete after the next "tick".